lib/engine/game/g_1822_pnw/step/merge.rb

```diff
@@ -0,0 +1,214 @@
+              n >= min_exchange_shares(par, minors_value,
+                                       minors_cash) && n <= max_exchange_shares(par, minors_value,
+                                                                                player_cash) && (6 * par) >= minors_value
+            end
+          end
+
+          def can_par_at?(par, minors_cash, minors_value, player_cash)
+            return true unless possible_exchanged_shares(par, minors_cash, minors_value, player_cash).empty?
+          end
+
+          def valid_par_prices(minor_one, minor_two)
+            minors_value = (minor_one.share_price.price + minor_two.share_price.price) * 2
+            minors_cash = minor_one.cash + minor_two.cash + 100
+            @game.stock_market.par_prices.map(&:price).sort.filter do |par|
+              can_par_at?(par, minors_cash, minors_value, minor_one.owner.cash)
+            end
+          end
+
+          def choices
+            choices = {}
+            case @merge_state
+            when :selecting_par
+              valid_par_prices(@associated_minor, @unassociated_minor).each do |par|
+                choices[par.to_s] = @game.format_currency(par)
+              end
+            when :selecting_shares
+              minors_value = (@associated_minor.share_price.price + @unassociated_minor.share_price.price) * 2
+              minors_cash = @associated_minor.cash + @unassociated_minor.cash + 100
+              player_cash = @associated_minor.owner.cash
+              possible_exchanged_shares(@selected_par, minors_cash, minors_value, player_cash).each do |shares|
+                money_difference = minors_value - (@selected_par * shares)
+                choices[shares.to_s] = if money_difference.zero?
+                                         "#{@associated_minor.owner.name} receives #{shares} shares"
+                                       elsif money_difference.positive?
+                                         "#{@associated_minor.owner.name} receives #{shares} shares and " \
+                                           "#{@game.format_currency(money_difference)}"
+                                       else
+                                         "#{@associated_minor.owner.name} receives #{shares} shares and pays " \
+                                           "#{@game.format_currency(-1 * money_difference)}"
+                                       end
+              end
+            when :selecting_token
+              choices['replace'] = 'Replace token on the map with an exchange token'
+              choices['exchange'] = 'Move an exchange token to available token area on the corporation chart'
+            end
+            choices
+          end
+
+          def process_merge(action)
+            @associated_minor = action.entity
+            @unassociated_minor = action.corporation
+
+            if !@unassociated_minor || !mergeable(@associated_minor).include?(@unassociated_minor)
```

Comment:
Put an @loading check here since you are using graph in merageable

Reply:
Done